libmogwai-schedule/tests/scheduler.c

```diff
@@ -0,0 +1,405 @@
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * Authors:
+ *  - Philip Withnall <withnall@endlessm.com>
+ */
+
+#include "config.h"
+
+#include <glib.h>
+#include <gio/gio.h>
+#include <libmogwai-schedule/scheduler.h>
+#include <libmogwai-schedule/tests/connection-monitor-dummy.h>
+#include <libmogwai-schedule/tests/peer-manager-dummy.h>
+#include <libmogwai-schedule/tests/signal-logger.h>
+#include <locale.h>
+
+
+/* Fixture which creates a #MwsScheduler, connects it to a mock peer manager,
+ * and logs all its signal emissions in @scheduled_signals.
+ *
+ * @scheduler_signals must be empty when the fixture is torn down. */
+typedef struct
+{
+  MwsConnectionMonitor *connection_monitor;  /* (owned) */
+  MwsPeerManager *peer_manager;  /* (owned) */
+  MwsScheduler *scheduler;  /* (owned) */
+  MwsSignalLogger *scheduler_signals;  /* (owned) */
+} Fixture;
+
+static void
+setup (Fixture       *fixture,
+       gconstpointer  test_data)
+{
+  fixture->connection_monitor = MWS_CONNECTION_MONITOR (mws_connection_monitor_dummy_new ());
+  fixture->peer_manager = MWS_PEER_MANAGER (mws_peer_manager_dummy_new (FALSE));
+
+  fixture->scheduler = mws_scheduler_new (fixture->connection_monitor,
+                                          fixture->peer_manager);
+  fixture->scheduler_signals = mws_signal_logger_new ();
+  mws_signal_logger_connect (fixture->scheduler_signals,
+                             fixture->scheduler, "notify::allow-downloads");
+  mws_signal_logger_connect (fixture->scheduler_signals,
+                             fixture->scheduler, "notify::entries");
+  mws_signal_logger_connect (fixture->scheduler_signals,
+                             fixture->scheduler, "entries-changed");
+  mws_signal_logger_connect (fixture->scheduler_signals,
+                             fixture->scheduler, "active-entries-changed");
+}
+
+static void
+teardown (Fixture       *fixture,
+          gconstpointer  test_data)
+{
+  g_clear_object (&fixture->scheduler);
+  g_clear_object (&fixture->peer_manager);
+  g_clear_object (&fixture->connection_monitor);
+
+  mws_signal_logger_assert_no_emissions (fixture->scheduler_signals);
+  g_clear_pointer (&fixture->scheduler_signals, (GDestroyNotify) mws_signal_logger_free);
+}
+
+/**
+ * assert_ptr_arrays_equal:
+ * @array1: (nullable): an array
+ * @array2: (nullable): another array
+ *
+ * Compare @array1 and @array2, checking that all their elements are pointerwise
+ * equal.
+ *
+ * Passing %NULL to either parameter is equivalent to passing an empty array.
+ *
+ * If the arrays do not match, an assertion will fail.
+ *
+ * Since: 0.1.0
+ */
+static void
+assert_ptr_arrays_equal (GPtrArray *array1,
+                         GPtrArray *array2)
+{
+  gboolean array1_is_empty = (array1 == NULL || array1->len == 0);
+  gboolean array2_is_empty = (array2 == NULL || array2->len == 0);
+
+  g_assert_cmpuint (array1_is_empty, ==, array2_is_empty);
```

Comment:
Shouldn't it rather be `g_assert_true` and just test if both vars are equal? I know in the end it's the same but this function doesn't seem the right one for comparing booleans.

Reply:
I deliberately chose `g_assert_cmpuint()` here because it prints out the values of the two operands on failure — `g_assert_true()` doesn’t. Ideally we’d have a `g_assert_cmpbool()` to do the same, but I haven’t got around to adding it (and I’m not sure it’s worth it).